src/inmanta/agent/agent.py

```diff
@@ -1691,30 +1693,91 @@ async def create_environment(self, blueprint: EnvBlueprint, threadpool: ThreadPo
         :param blueprint: The blueprint specifying the configuration for the new virtual environment.
         :param threadpool: A ThreadPoolExecutor
         :return: An instance of ExecutorVirtualEnvironment representing the created or reused environment.
-
-        TODO: Improve handling of bad venv scenarios, such as when the folder exists but is empty or corrupted.
         """
         env_storage, is_new = self.get_or_create_env_directory(blueprint)
+
+        # Check for validity of the existing virtual environment
+        if not is_new and not self.is_environment_valid(env_storage, blueprint):
+            LOGGER.info("Existing virtual environment at %s is invalid or corrupted and will be recreated", env_storage)
+            self.recreate_environment_directory(env_storage)
+            is_new = True
+
         process_environment = ExecutorVirtualEnvironment(env_storage, threadpool)
         if is_new:
             await process_environment.create_and_install_environment(blueprint)
         self._environment_map[blueprint] = process_environment
 
         return process_environment
 
+    def is_environment_valid(self, env_storage: str, blueprint: EnvBlueprint) -> bool:
+        """
+        Checks if the given virtual environment directory is valid by looking for specific files and directories
+        that should exist in a properly configured virtual environment.
+
+        :param env_storage: The file system path to the virtual environment.
+        :param blueprint: The blueprint specifying the configuration for the new virtual environment.
+        :return: True if the environment is valid, False otherwise.
+        """
+        expected_elements = ["pyvenv.cfg", "lib", "bin"]
+        try:
+            if not all(os.path.exists(os.path.join(env_storage, element)) for element in expected_elements):
+                return False
+            installed_packages = self.get_installed_packages(env_storage)
+            required_packages = {pkg.split("==")[0] for pkg in blueprint.requirements}
+            missing_packages = required_packages - set(installed_packages.keys())
+        except Exception as e:
+            LOGGER.info(f"Error while validating environment: {e}")
+            return False
+        if missing_packages:
+            return False
+        return True
+
+    def get_installed_packages(self, env_storage: str) -> dict:
+        """
+        Gets a dictionary of installed packages and their versions
+        :param env_storage: The file system path to the virtual environment.
+        :return: A dictionary with package names as keys and their versions as values.
+        """
+        pip_path = os.path.join(env_storage, "bin", "pip")
```

Comment:
I would prefer not to fork pip.
We are going to rather extreme lengths to make this process efficient, forking pip up to 3 times in a single `get_environment` operation is not an option (not even once, tbh)

Reply:
@wouterdb than I'll close this PR/Branch and put the ticket back